File: leetcode/15.py
class Solution:
  def threeSum(self, nums: list[int]) -> list[list[int]]:
    if len(nums) < 3:
      return []

    ans = []
    nums.sort()

    for i in range(len(nums) - 2):
      if nums[i] > 0:
        break

      if i > 0 and nums[i] == nums[i - 1]:
        continue
  
      l, r = i + 1, len(nums) - 1
      
      while l < r:
        summ = nums[i] + nums[l] + nums[r]
        if summ == 0:
          ans.append([nums[i], nums[l], nums[r]])
          
          l += 1
          while l < r and nums[l] == nums[l - 1]:
            l += 1
          
          r -= 1
          while l < r and nums[r] == nums[r + 1]:
            r -= 1

        elif summ < 0:
          l += 1
        else:
          r -= 1

    return ans

File: leetcode/15.cpp
class Solution {
public:
  vector<vector<int>> threeSum(vector<int> &nums) {
    if (nums.size() < 3)
      return {};

    vector<vector<int>> ans;

    ranges::sort(nums);

    for (int i = 0; i + 2 < nums.size(); ++i) {
      if (i > 0 && nums[i] == nums[i - 1])
        continue;

      int left = i + 1;
      int right = nums.size() - 1;

      while (left < right) {
        const int sum = nums[i] + nums[left] + nums[right];

        if (sum == 0) {
          ans.push_back({nums[i], nums[left++], nums[right--]});

          while (left < right && nums[left] == nums[left - 1])
            ++left;

          while (left < right && nums[right] == nums[right + 1])
            --right;
        } else if (sum < 0) {
          ++left;
        } else {
          --right;
        }
      }
    }

    return ans;
  }
};